UnitTests/CommonTestUtils/SubmoduleTestHelpers.cs

```diff
@@ -0,0 +1,40 @@
+﻿using System.Threading;
+using GitCommands;
+using GitCommands.Submodules;
+
+namespace CommonTestUtils
+{
+    public class SubmoduleTestHelpers
+    {
+        public static SubmoduleInfoResult UpdateSubmoduleStatusAndWaitForResult(ISubmoduleStatusProvider provider, GitModule module, bool updateStatus = false)
+        {
+            SubmoduleInfoResult result = null;
+            SemaphoreSlim onUpdateCompleteSignal = new SemaphoreSlim(0, 1);
+
+            provider.StatusUpdating += Provider_StatusUpdating;
+            provider.StatusUpdated += Provider_StatusUpdated;
+
+            provider.UpdateSubmodulesStatus(
```

Comment:
I started refactoring this PR and found out that there are other hidden async functions. One of the reasons, as I wrote before, is to be able to await an async method. This is an example where the async nature is hidden and instead of just awaiting:
```csharp
await provider.UpdateSubmodulesStatusAsync(
```
some backdoor is used to get notified that `UpdateSubmodulesStatus` is completed. If there is a bug inside of the `UpdateSubmodulesStatus` and the Provider_StatusUpdated is not called, this test will never end.

I believe there are places in the current code base where the async nature is hidden, but they were made this way only to not to have to migrate the old code to full async. In new code I would expect to have full async code.

Reply:
> If there is a bug inside of the `UpdateSubmodulesStatus` and the Provider_StatusUpdated is not called, this test will never end.

💭 This may explain why FormBrowse tests are randomly hang...

Reply:
Just to be clear, you want me to make UpdateSubmoduleStatus async, right? But I still need it to call back multiple listeners added to StatusUpdating and StatusUpdated events for FormBrowse and SubmoduleTree. That's alright?

Reply:
>Just to be clear, you want me to make UpdateSubmoduleStatus async, right?

Yes.

>But I still need it to call back multiple listeners added to StatusUpdating and StatusUpdated events for FormBrowse and SubmoduleTree. That's alright?

Yes, and both StatusUpdating and StatusUpdated should be async as well.

Then only `FormBrowse.UpdateSubmodulesStructure` will need to wrap `UpdateSubmoduleStatusAsync` into `JTF.RunAsync`.

I think that the rule of thumb might be that whenever you attach `FileAndForget` at the of anything but `JTF.RunAsync` it is a sign that sth is not designed along with the JTF.

Reply:
> Yes, and both StatusUpdating and StatusUpdated should be async as well.

I haven't done async event handlers yet. Would I need to do something as shown in this answer: https://stackoverflow.com/a/27763068/4039972
?

Reply:
Good question! I have never done it too. My proposal would be:

`AsyncEventArgs `

```csharp

    public class AsyncEventArgs : EventArgs
    {
        public List<Task> Tasks { get; } = new List<Task>();
    }
```
`SubmoduleStatusProvider`

```csharp
        private async Task OnStatusUpdatedAsync(SubmoduleInfoResult info, CancellationToken token)
        {
            await TaskScheduler.Default;
            token.ThrowIfCancellationRequested();
            _lastStatusEventArgs = new SubmoduleStatusEventArgs(info, token);
            StatusUpdated?.Invoke(this, _lastStatusEventArgs);
            await Task.WhenAll(_lastStatusEventArgs.Tasks);
        }
```
`FormBrowse`

```csharp
        private void SubmoduleStatusProvider_StatusUpdated(object sender, SubmoduleStatusEventArgs e)
        {
           e.Tasks.Add(PopulateToolbarAsync(e.Info, e.Token));
        }
```

@sharwell Do you have experience in that field?

Reply:
@jbialobr Is the goal here just to allow waiting for asynchronous background operations to complete in UI tests? If so, you can just `await ThreadHelper.JoinPendingOperationsAsync()` in your test code (DO NOT USE IN PRODUCTION CODE).

Reply:
Yes, it was my main motivation. How does `
JoinPendingOperationsAsync` work? Does it wait for all JTF.RunAsync calls or only for those which were continued with FileAndForget? Does it throw OperationCanceledException if any of the awaited tasks was cancelled (by throwing OperationCanceledException)?